### PR TITLE
rgw/cloudtier: initial MVP of the feature RestoreObject from cloud 

### PR DESCRIPTION
**Pull request by soumyakoduri**
This PR covers initial PoC for the feature to restore objects from the cloud-tier on demand. Details below -

Feature doc: https://docs.google.com/document/d/1rzLJAzHK6cLuzJswgoplgugNOFCKi8NrjPOGfgVrIFg/edit

**Restore CLI parsing**
Added restore cli parameters and xml parsing.
Current commit has basic cli checks with their corresponding response, cli parameters like Days, OutputLocation( storage class, target bucket name , prefix)

TODO:
- [ ] Fetching head object and reading storage-class parameters.
- [ ] Restore in  progress cli.
- [ ] Improve response messages.
Signed-off-by: Shreyansh Sancheti <ssancheti@ibm.com>

**GET/RESTORE** : calling cloud restore apis in get and restore call
GET request- It allows read-through for cloud-tiered objects via restore_obj_from_cloud
RESTORE request - It checks whether object is cloud tiered, returns error if not otherwise fetch object via restore_obj_from_cloud

TODO:
- [ ]    add option to allow cloud_restore option similar to retain_head_object
- [ ]     Implementation of callbacks
- [ ]     Even read through works, for the first request user gets empty object
- [ ]     Make sure all the values are passed correctly to restore_obj_from_cloud
- [ ]     Parsed parameters like o/p location, days in restore call pass to down to restore_obj_from_cloud
- [ ]     Include cli option for restore days incase of get request
- [ ]     Handling restore status
Signed-off-by: Jiffin Tony Thottan [thottanjiffin@gmail.com](mailto:thottanjiffin@gmail.com)

**CloudTier module:**
* Given `<bucket,object,cloudtier-storage-class>` & `optional<days>`, fetch the object from the cloud endpoint and copy it to STANDARD storage-class
* if `<days>` provided,  and
                    a) > 0, the restore is marked temporary with expiry date, but
                    b) = 0, restore is not done and an error is returned
* Without `<days>`, it is considered as permanent restore.
*  For temporary restore, return original cloudtier storage-class for `x-amz-storage-class` .

TODO:
- [ ] ObjectExpirer to delete temporary restored objects
- [ ] handle LC part
- [ ] s3cmd attrs and any other attr related changes
Signed-off-by: Soumya Koduri <skoduri@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -3408,6 +3408,168 @@ int RGWPostObj_ObjStore_S3::get_encrypt_filter(
   return res;
 }
 
+struct S3Location {
```
Comment (soumyakoduri) [inline src/rgw/rgw_rest_s3.cc:3411]: @shreyanshjain7174 .. as discussed with Kyle and Daniel in the slack, we are not doing OutputLocation as it is associated with Restore Select query.

**Review by thotz — CHANGES_REQUESTED** [review]
Reviewed only Shreyansh commit

File: src/rgw/rgw_op.cc [hunk 2]
```diff
@@ -5150,6 +5180,77 @@ void RGWPutMetadataObject::execute(optional_yield y)
   op_ret = s->object->set_obj_attrs(this, &attrs, &rmattrs, s->yield, rgw::sal::FLAG_LOG_OP);
 }
 
+int RGWRestoreObj::init_processing(optional_yield y)
+{
+  /* Read in the data from the POST form. */
+  int ret = get_params(y);
```
[the remark this answers is not in the document]
Comment (shreyanshjain7174) [reply src/rgw/rgw_op.cc:5186]: Got it! Thanks

File: src/rgw/rgw_op.cc [hunk 3]
```diff
@@ -5150,6 +5180,77 @@ void RGWPutMetadataObject::execute(optional_yield y)
   op_ret = s->object->set_obj_attrs(this, &attrs, &rmattrs, s->yield, rgw::sal::FLAG_LOG_OP);
 }
 
+int RGWRestoreObj::init_processing(optional_yield y)
+{
+  /* Read in the data from the POST form. */
```
[the remark this answers is not in the document]
Comment (shreyanshjain7174) [reply src/rgw/rgw_op.cc:5185]: Not exactly, I will remove it.

File: src/rgw/rgw_op.cc [hunk 4]
```diff
@@ -5150,6 +5180,77 @@ void RGWPutMetadataObject::execute(optional_yield y)
...
+      op_ret = handle_cloudtier_obj(s, this, driver, attrs, false, tier_config, y);
+      if (op_ret < 0) {
+        ldpp_dout(this, 4) << "Cannot get cloud tiered object: " << *s->object
+        <<". Failing with " << op_ret << dendl;
+        if (op_ret == -ERR_INVALID_OBJECT_STATE) {
+          s->err.message = "This object was transitioned to cloud-s3";
+        }
+      }
+    } else {
+      ldpp_dout(this, 20) << "not cloud tier object erroring" << dendl;
```
[the remark this answers is not in the document]
Comment (shreyanshjain7174) [reply src/rgw/rgw_op.cc:5243]: @soumyakoduri what do you say?

File: src/rgw/rgw_rest.cc [hunk 5]
```diff
@@ -1458,6 +1458,37 @@ int RGWPostObj_ObjStore::get_params(optional_yield y)
   return 0;
 }
 
+int RGWRestoreObj_ObjStore::verify_params()
+{
```
[the remark this answers is not in the document]
Comment (shreyanshjain7174) [reply src/rgw/rgw_rest.cc:1462]: I think we can skip verify_params, I copied from other methods I guess, but to think now we don't have a use of it. As for days we will have to check it for cli in rgw_rest_s3 and separately for tier config related check.

File: src/rgw/rgw_rest_s3.cc [hunk 6]
```diff
@@ -3435,6 +3451,168 @@ int RGWPostObj_ObjStore_S3::get_encrypt_filter(
...
+  if (op_ret < 0)
+  {
+    return op_ret;
+  }
+
+  std::string bucket_name;
+  std::string storage_class;
+  std::string prefix;
+  std::string expected_bucket_owner;
+
```
[the remark this answers is not in the document]
Comment (shreyanshjain7174) [reply src/rgw/rgw_rest_s3.cc:3526]: Yeah I will remove those.

File: src/rgw/rgw_rest.cc [hunk 7]
```diff
@@ -1458,6 +1458,37 @@ int RGWPostObj_ObjStore::get_params(optional_yield y)
...
+  return 0;
+}
+
+int RGWRestoreObj_ObjStore::get_params(optional_yield y)
+{ 
+  string bucket_name;
+  string object_name;
+  string version_id;
+  // wildcard can also be given such as '*'
+  if (!s->bucket->get_name().empty()) {
```
[the remark this answers is not in the document]
Comment (shreyanshjain7174) [reply src/rgw/rgw_rest.cc:1480]: ©I think I will remove `get_params` from here its not needed for the admin ops cli. I am already using `read_all_inputs` in the `rest_s3`.

File: src/rgw/driver/rados/rgw_sal_rados.cc [hunk 8]
```diff
@@ -2551,6 +2552,93 @@ int RadosObject::transition(Bucket* bucket,
...
+  tier_ctx.acl_mappings = rtier->get_rt().t.s3.acl_mappings;
+  tier_ctx.multipart_min_part_size = rtier->get_rt().t.s3.multipart_min_part_size;
+  tier_ctx.multipart_sync_threshold = rtier->get_rt().t.s3.multipart_sync_threshold;
+  tier_ctx.storage_class = tier->get_storage_class();
+
+  ldpp_dout(dpp, 20) << "Restoring object(" << o.key << ") from the cloud endpoint(" << endpoint << ")" << dendl;
+
+  if (days && days == 0) {
+    ldpp_dout(dpp, 0) << "Days = 0 not valid; Not restoring object (" << o.key << ") from the cloud endpoint(" << endpoint << ")" << dendl;
+    return -1;
```
Comment (thotz) [inline src/rgw/driver/rados/rgw_sal_rados.cc:2600]: returns ENVALID??

File: src/rgw/driver/rados/rgw_rados.cc [hunk 9]
```diff
@@ -5123,6 +5123,170 @@ int RGWRados::transition_obj(RGWObjectCtx& obj_ctx,
...
+                             RGWObjectCtx& obj_ctx,
+                             RGWBucketInfo& dest_bucket_info,
+                             const rgw_obj& dest_obj,
+                             rgw_placement_rule& dest_placement,
+                             real_time& mtime,
+                             uint64_t olh_epoch,
+                             std::optional<uint64_t> days,
+                             const DoutPrefixProvider *dpp,
+                             optional_yield y,
+                             bool log_op){
```
Comment (thotz) [inline src/rgw/driver/rados/rgw_rados.cc:5136]: unused variable

File: src/rgw/driver/rados/rgw_lc_tier.cc [hunk 10]
```diff
@@ -261,20 +282,78 @@ static int cloud_tier_get_object(RGWLCCloudTierCtx& tier_ctx, bool head,
...
-    ldpp_dout(tier_ctx.dpp, 0) << "ERROR: " << __func__ << "(): conn.get_obj() returned ret=" << ret << dendl;
-    return ret;
+  ldpp_dout(tier_ctx.dpp, 20) << __func__ << "(): fetching object from cloud bucket:" << dest_bucket << ", object: " << target_obj_name << dendl;
+
+  static constexpr int NUM_ENPOINT_IOERROR_RETRIES = 20;
+  for (int tries = 0; tries < NUM_ENPOINT_IOERROR_RETRIES; tries++) {
+    ret = tier_ctx.conn.get_obj(tier_ctx.dpp, dest_obj, req_params, true /* send */, &in_req);
+    if (ret < 0) {
+      ldpp_dout(tier_ctx.dpp, 0) << "ERROR: " << __func__ << "(): conn.get_obj() returned ret=" << ret << dendl;
+      return ret;
```
Comment (thotz) [inline src/rgw/driver/rados/rgw_lc_tier.cc:294]: why continue retry logic is not used here like in Linenumbber 305

**Review by thotz — CHANGES_REQUESTED** [review]

File: src/rgw/rgw_rest_s3.cc (hunk 1 above)
Comment (shreyanshjain7174) [reply src/rgw/rgw_rest_s3.cc:3435]: okay! done

File: src/rgw/rgw_lc.cc [hunk 11]
```diff
@@ -1404,7 +1404,39 @@ class LCOpAction_Transition : public LCOpAction {
     if (obj->get_attr(RGW_ATTR_ETAG, bl)) {
       etag = rgw_bl_str(bl);
     }
-    auto size = obj->get_size();
```
Comment (shreyanshjain7174) [inline src/rgw/rgw_lc.cc:1407]: size is not initalised locally for this function to use for send_notification()

File: src/rgw/driver/rados/rgw_sal_rados.cc [hunk 12]
```diff
@@ -2628,6 +2716,23 @@ int RadosObject::transition_to_cloud(Bucket* bucket,
...
 
+int RadosObject::set_cloud_restore_status(const DoutPrefixProvider* dpp,
+				  optional_yield y,
+          rgw::sal::RGWRestoreStatus restore_status)
+{
+  int ret = 0;
+  set_atomic();
+ 
+  bufferlist bl;
+  //bl2.append("CLOUD_RESTORE_IN_PROGRESS");
```
Comment (thotz) [inline src/rgw/driver/rados/rgw_sal_rados.cc:2727]: unintended??

File: src/rgw/rgw_lc.cc [hunk 13]
```diff
@@ -1406,6 +1406,15 @@ class LCOpAction_Transition : public LCOpAction {
     }
     auto size = obj->get_size();
 
+    if (obj->get_attr(RGW_ATTR_RESTORE_STATUS, bl)) {
+      std::string restore_status = rgw_bl_str(bl);
+      if (restore_status == "CLOUD_RESTORED") {
```
Comment (shreyanshjain7174) [inline src/rgw/rgw_lc.cc:1411]: should be CloudRestored @soumyakoduri ?

**Pull request closed without merging**